src/third_party/GMSHParser.cpp

```diff
@@ -0,0 +1,214 @@
+#include "GMSHParser.h"
+#include "GMSHLexer.h"
+
+#include <cstdio>
+#include <fstream>
+#include <sstream>
+
+namespace tndm {
+
+template <typename T> T GMSHParser::logErrorAnnotated(std::string_view msg) {
+    std::stringstream ss;
+    ss << "GMSH parser error in line " << curLoc.line << " in column " << curLoc.col << ":\n";
+    ss << '\t' << msg << '\n';
+    errorMsg += ss.str();
+    return {};
+}
+
+template <typename T> T GMSHParser::logError(std::string_view msg) {
+    errorMsg += "GMSH parser error:\n\t";
+    errorMsg += msg;
+    errorMsg += '\n';
+    return {};
+}
```

Comment:
This confuses me - why is there a template here? And how does it actually work in a .cpp file?

Reply:
There's a template such that you can match the return type of the function. E.g.
return logErrorAnnotated<double>("Expected $MeshFormat");
return logErrorAnnotated<bool>("Expected non-zero integer");

It works within a cpp file because the function is local to the cpp file.